subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt_util.c

```diff
@@ -16,8 +16,9 @@ img_mgmt_ver_str(const struct image_version *ver, char *dst)
 	int rc = 0;
 	int rc1 = 0;
 
-	rc = snprintf(dst, IMG_MGMT_VER_MAX_STR_LEN, "%hhu.%hhu.%hu",
-		ver->iv_major, ver->iv_minor, ver->iv_revision);
+	rc = snprintf(dst, IMG_MGMT_VER_MAX_STR_LEN, "%hu.%hu.%hu",
+		(uint16_t)ver->iv_major, (uint16_t)ver->iv_minor,
+		ver->iv_revision);
```

Comment:
> The "New C Library" (CONFIG_NEWLIB_LIBC=y) does not correctly
process %hhu formatting, causing a version string to be returned
malformed

Could you elaborate? How does it get malformed? Also what toolchain/version did you use?

p.s. "New C Lib" is an interesting way to refer to the "newlib;" it should be just called "newlib" like a proper noun.

Reply:
> 
> Could you elaborate? How does it get malformed? Also what toolchain/version did you use?

Here is bug that describes the issue: https://github.com/zephyrproject-rtos/zephyr/issues/45261, in short:
when you do the "%hhu", to format char or uint8_t,  it will print "hu", for example:
```
$$ sudo ~/go/bin/mcumgr -t 1200 --conntype ble --connstring ctlr_name=hci0,peer_name='Zephyr' image list
Images:
 image=0 slot=0
    version: 0.0.0
    bootable: true
    flags: active confirmed
    hash: 8fc300ef78d8471420c9f46e7ff311707c48807c253df48b54a847c8e5db7fd7
 image=0 slot=1
    version: 0.0.0
    bootable: true
    flags:
    hash: e3364341a7fc9ecaada6709b757ba72e90389012eaea7528999c9d39afd01ee1
Split status: N/A (0)
```
The SDK is zephyr-sdk-0.14.1, the SHA of Zephyr is d5e7732e0d0f66bc9a3684f9597d9fd6a2d351b2
> 
> p.s. "New C Lib" is an interesting way to refer to the "newlib;" it should be just called "newlib" like a proper noun.

Yeah, that was me being dumb and typing the "Kconfig" option prompt for memory, which obviously should be "Newlib C library"

I usually build and test the mcumgr,  by the smp_svr sample, on newlib and the minimal one, but somehow failed to do this this time and have failed to notice the problem.

Reply:
Ok, I just tested this and `%hhu` and its friends are indeed broken (not enabled) -- moreover, `PRIu8` resolves to `%hhu` even for the newlib, so this is indeed very broken.

A brief investigation suggests that our newlib is being built without `_WANT_IO_C99_FORMATS`, so the `hh` modifier support is not enabled:
https://github.com/zephyrproject-rtos/newlib-cygwin/blob/2893dbb10f1fe3bf431a26f89a22e12a5020dc02/newlib/libc/stdio/vfprintf.c#L1105-L1110

I will file a separate bug for this and eventually get this fixed in the Zephyr SDK.

Reply:
UPDATE: `_WANT_IO_C99_FORMATS` is enabled for the full newlib build, so it should work if you build with `CONFIG_NEWLIB_LIBC_NANO=n` (nano variant is used by default on ARM).

Nevertheless, the fact that `PRIu8` is `%hhu` for the newlib nano in spite of it not supporting the `hh` modifiers is wrong and needs to be fixed.

Reply:
> 
> A brief investigation suggests that our newlib is being built without `_WANT_IO_C99_FORMATS`, so the `hh` modifier support is not enabled: https://github.com/zephyrproject-rtos/newlib-cygwin/blob/2893dbb10f1fe3bf431a26f89a22e12a5020dc02/newlib/libc/stdio/vfprintf.c#L1105-L1110

Quick find  👍 

> 
> I will file a separate bug for this and eventually get this fixed in the Zephyr SDK.

Thanks.

Reply:
> UPDATE: `_WANT_IO_C99_FORMATS` is enabled for the full newlib build, so it should work if you build with `CONFIG_NEWLIB_LIBC_NANO=n` (nano variant is used by default on ARM).
> 

@stephanosio  Maybe it is not worth enabling another part of newlib  for this. I will just change commit message to not blame newlib on that and leave the fix.

Reply:
> @stephanosio Maybe it is not worth enabling another part of newlib for this. I will just change commit message to not blame newlib on that and leave the fix.

The fundamental issue is that `inttypes.h` is not correctly resolving (I think it is a header inclusion issue, that the newlib-nano `newlib.h` is not included, and instead the full newlib `newlib.h` is included by `inttypes.h`) the `PRIu8` inttype.

In case of the newlib-nano, `PRIu8` should be `%hu` (the closest type to `%hhu`), in which case you should be able to generalise the patch as `PRIu8 "." PRIu8 "." PRIu16`.

Reply:
Filed #45336

Reply:
> Filed #45336

Thank you!